### DISCUSSION TITLE
Smart Meter Variable assignment "v1=SML#VoltageL1" does not work

### DISCUSSION THREAD

**Discussion by MichaelUray** [arendst/Tasmota#15339] [Q&A] [answered]
Hi guys,

I try to read smart meter data into a variable, but for some reason it does not work.
The data itself is shown on the website, what means the reading is ok, but I am not able to assign the data to a variable to do further calculations with it.

The [documentation](https://tasmota.github.io/docs/Smart-Meter-Interface) shows two different ways in the examples, but none of them work.
`v1=SML#VoltageL1` - there the value does not change (stays on its init value 222)
and
`v1=#VoltageL1` - there the value gets overwritten with 0.

If I try `v1=Wifi#RSSI`, then it works fine and I get the RSSI value.

This is my script:
```
>D
scnt=0
res=0
v1=222

;TelePeriod time
tper=15

>B
=>sensor53 r

>F
; count 100ms
scnt+=1
switch scnt
case 2
;set sml driver to 300 baud and send /?! as HEX to trigger the Meter
res=sml(1 0 300)
res=sml(1 1 "2F3F210D0A")

;1500ms later \> Ack and ask for switching to 9600 baud
case 17
res=sml(1 1 "063035300D0A")

;300ms later \> Switching sml driver to 9600 baud
case 20
res=sml(1 0 9600)

;restart reading at end of teleperiod
if scnt>90
and upsecs%tper==0
then
scnt=0
endif

>T
v1=SML#VoltageL1

>J
,"Voltage":%2v1%

>M 1
+1,3,o,0,9600,,1
1,32.7(@1,VoltageL1,V,VoltageL1,1
```

This is my `user_config_override.h`, tasmota-lite is chosen.
```
#ifdef USE_RULES
#undef USE_RULES
#endif

#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif

#define USE_SCRIPT_SUB_COMMAND
#define USE_SCRIPT_GLOBVARS

#define USE_SCRIPT_JSON_EXPORT
#define USE_ANGLE_FUNC

#define SCRIPT_LARGE_VNBUFF

#ifndef USE_SML_M
#define USE_SML_M
#endif

#define SML_REPLACE_VARS
#define USE_SML_SCRIPT_CMD

#undef SML_MAX_VARS
#define SML_MAX_VARS 25
#undef MAX_METERS
#define MAX_METERS  1


#ifndef USE_MQTT_TLS
#define USE_MQTT_TLS                             // Use TLS for MQTT connection (+34.5k code, +7.0k mem and +4.8k additional during connection handshake)
 #define USE_MQTT_TLS_CA_CERT                   // Force full CA validation instead of fingerprints, slower, but simpler to use.  (+2.2k code, +1.9k mem during connection handshake)
                                                   // This includes the LetsEncrypt CA in tasmota_ca.ino for verifying server certificates
//  #define USE_MQTT_TLS_FORCE_EC_CIPHER           // Force Elliptic Curve cipher (higher security) required by some servers (automatically enabled with USE_MQTT_AWS_IOT) (+11.4k code, +0.4k mem)
#endif

//#undef MQTT_FINGERPRINT1
//#define MQTT_FINGERPRINT1     "FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF"

#undef SetOption132
#define SetOption132 0

#define USE_TCP_BRIDGE

#ifdef USE_TIMERS
#undef USE_TIMERS                              
#endif 
```

Any idea what I do wrong there?

**Comment by gemu2015** [comment, marked as answer]
you did not specify any topic for your meter.
\>M 1
+1,3,o,0,9600,,1

in case you want to call it SML you must define
\>M 1
+1,3,o,0,9600,SML,1

you may also use sml[x] x being the decode line (1 .. N)
since you can use this variable throughout the script, not only in >T

**Reply by MichaelUray** [reply]
> in case you want to call it SML you must define
> >M 1
> +1,3,o,0,9600,SML,1
Oh boy, it is that easy? I did look for hours to find the problem and I did not see it.
Thank you very much!

**Comment by simsasaile** [comment]
Hi guys,

I have the same symptom, I can't assign the values to variables, the console just shows ??? instead of the value. But the cause seems to be different, does anyone have an idea?

This is my script:

```
>D
energy = 0

>B
=>sensor53 r

>T
energy=meter#energy

>S
if upsecs%5==0
then
print Energy: %energy%

>M 1
+1,3,s,16,9600,meter,1
1,77070100010800ff@1000,Energie,kWh,energy,4
1,77070100020800ff@1000,Lieferung,kWh,en_out,4
1,77070100100700ff@1,Leistung,W,power,2
1,7707010060320101@#,SID,,meter_id,0
#
```

**Reply by simsasaile** [reply]
Replacing meter#energy with sml[1] and using permanent vars solved it. Still don't know why this way doesn't work, maybe some requirements have to be include into the builds?

**Reply by gemu2015** [reply]
no spaces allowed in declaration.  (energy = 0)

energy=0 

will work
